docs: add ai form filler articles (#5628)
vaadin/docs: articles/flow/ai-support/ai-powered-form.adoc

FormLayout form = new FormLayout(name, email, country, hiredOn);
MessageInput messageInput = new MessageInput();

FormAIController controller = new FormAIController(form);

AIOrchestrator.builder(provider, systemPrompt)
        .withInput(messageInput)
        .withController(controller)
        .build();

add(messageInput, form);
----

Example prompts:

* "Fill in John Doe, john@acme.com, started in Germany on 2026-03-01."
* "Maria from Helsinki, hired last Monday." (relative dates are interpreted by the LLM)
* "Use the information in the attached resume to fill the form." (when the orchestrator has a <<file-attachments#,file receiver>> configured)

.Built-In Workflow Instructions
[TIP]
The controller already informs the LLM of the workflow it needs. You can focus your own system prompt on application-specific behavior, such as tone, naming conventions, or which fields the user may leave blank.

== Field Discovery

The controller walks the container's component tree on every LLM turn, so fields added or removed between turns are picked up automatically. Any component that implements [classname]`HasValue` is treated as a field; nested layouts are walked recursively.

> Each field gets an opaque identifier that the controller uses on the LLM side. The identifier is stored on the field component, so it stays the same when a field is detached and re-attached within a session.

Reviewer: This is an internal technical detail and doesn't bring extra value to readers. I'd suggest dropping it from here.

Author: Removed the last part.

[source,java]
----
controller.ignore(internalIdField);
----

Ignored fields are hidden from the LLM and stay editable while a fill is in progress.

== Field Descriptions

The LLM sees each field's label, helper text, and component type. When those don't fully capture the field's meaning, for example a numeric field that takes a percentage rather than an absolute amount, or a date that means "renewal date" rather than "purchase date", add an explicit description with [methodname]`describe()`:

[source,java]
----
controller.describe(discount, "Discount as a percentage between 0 and 100.")
        .describe(renewalDate, "When the subscription renews, not when it started.");
----

Later calls for the same field overwrite earlier ones.

=== Binder Integration

When the form is backed by a [classname]`Binder`, pass the binder to the controller as well:

[source,java]
----
Binder<Employee> binder = new Binder<>(Employee.class);
binder.bindInstanceFields(this);

FormAIController controller = new FormAIController(form, binder);
----

> For every named binding (`bind("propertyName")`, `bindInstanceFields(this)`, or `@PropertyId`), the bean property name is used as a default field description, so the LLM can refer to the field by its bean-side name. An explicit [methodname]`describe()` call always overrides the default. Lambda-bound bindings have no property name and contribute no default.

Reviewer: > LLM can refer to the field by its bean-side name.

LLM only refers to the fields by their [field id](https://github.com/vaadin/flow-components/blob/f10450b798b04939b1cb08d77f3417daa6733398/vaadin-ai-components-flow-parent/vaadin-ai-components-flow/src/main/java/com/vaadin/flow/component/ai/form/FormAIController.java#L141), right?

Author: Reworded the last part of the sentenced to clarify the relationship.

---

> title: AI-Powered Form

Reviewer 2: ```suggestion
title: AI Form Filler
```

description: Use FormAIController to let users fill any Vaadin form from natural-language input or attached files.
meta-description: Configure the Vaadin FormAIController to fill form fields from natural language or attached files, with Binder validation and select-options support.
order: 80
---

= [since:com.vaadin:vaadin@V25.2]#AI-Powered Form#

> [classname]`FormAIController` populates the fields of a <<{articles}/components/form-layout#,[classname]`FormLayout`>> (Vaadin's responsive multi-column form container) or any other layout, using values an LLM extracts from a user prompt or attached files. The controller walks the layout, discovers every [classname]`HasValue` field, and lets the LLM read the current values, query select options for combo boxes and selects, and write new values back. Each write is validated through the [classname]`Binder`, or through the component's built-in validators when the field is not bound. Rejected values are reported back so the model can correct them on the next turn.

Reviewer 2: ```suggestion
[classname]`FormAIController` populates the fields of a <<{articles}/components/form-layout#,[classname]`FormLayout`>> (Vaadin's responsive multi-column form container) or any other layout, using values an LLM extracts from a user prompt or attached files. The controller traverses the layout, discovers every field, and allows the LLM to read the current values, query the available values for selection components like Combo Box or Radio Button Group, and write new values back. Each write is validated through the [classname]`Binder`, or through the component's built-in validators if [classname]`Binder` is not used. Rejected values are reported back so the model can correct them on the next turn.
```

[source,java]
----
TextField name = new TextField("Name");
EmailField email = new EmailField("Email");
ComboBox<String> country = new ComboBox<>("Country");
country.setItems("Finland", "Germany", "United States");
DatePicker hiredOn = new DatePicker("Hired on");

FormLayout form = new FormLayout(name, email, country, hiredOn);
MessageInput messageInput = new MessageInput();

FormAIController controller = new FormAIController(form);

AIOrchestrator.builder(provider, systemPrompt)
        .withInput(messageInput)
        .withController(controller)
        .build();

add(messageInput, form);
----

Example prompts:

* "Fill in John Doe, john@acme.com, started in Germany on March 1, 2026."
* "Maria from Helsinki, hired last Monday." (relative dates are interpreted by the LLM)
* "Use the information in the attached resume to fill the form." (when the orchestrator has a <<file-attachments#,file receiver>> configured)

.Built-In Workflow Instructions
[TIP]
The controller already informs the LLM of the workflow it needs. You can focus your own system prompt on application-specific behavior, such as tone, naming conventions, or which fields the user may leave blank.

> == Field Discovery

Reviewer 2: Should this section say something about what container elements can be used? (The docs talk about "layouts" which is kind of correct, but may give the impression that only layout components like HL, VL and FL are supported, when in fact it's any `HasComponents`, right?)

Author: Added clarification regarding the `HasComponents` requirement.

> The controller walks the container's component tree on every LLM turn, so fields added or removed between turns are picked up automatically. Any component that implements [classname]`HasValue` is treated as a field; nested layouts are walked recursively.

Reviewer 2: ```suggestion
The controller walks the container's component tree on every LLM turn, so fields added or removed between turns are picked up automatically. Any component that implements the [classname]`HasValue` interface is treated as a field; nested layouts are walked recursively.
```

[classname]`PasswordField` is always hidden from the LLM. To hide other fields, for example internal IDs or anything sensitive that the user must fill in manually, call [methodname]`ignore()`:

[source,java]
----
controller.ignore(internalIdField);
----

Ignored fields are hidden from the LLM and stay editable while a fill is in progress.

== Hidden, Disabled, and Read-Only Fields

The controller checks each field's state on every turn:

* A field hidden via [methodname]`setVisible(false)`, or one that sits inside a hidden container, is dropped from the LLM surface entirely. The model cannot read its value and cannot write to it. It reappears the moment a value-change listener or other application code makes it visible again.

> * A field the application has disabled ([methodname]`setEnabled(false)`) or set read-only ([methodname]`setReadOnly(true)`) stays in the form state as read-only context. The model sees its current value -- useful as context for writes to other fields -- but any write is rejected with a reason.

Reviewer 2: What does "with a reason" mean in the phrase "but any write is rejected with a reason."?

Author: The reason here means the rejection reason to be fed back to the LLM. Added clarification.

== Field Descriptions

The LLM sees each field's label, helper text, and component type. When those don't fully capture the field's meaning, for example a numeric field that takes a percentage rather than an absolute amount, or a date that means "renewal date" rather than "purchase date", add an explicit description with [methodname]`describe()`:

[source,java]
----
controller.describe(discount, "Discount as a percentage between 0 and 100.")
        .describe(renewalDate, "When the subscription renews, not when it started.");
----

Later calls for the same field overwrite earlier ones.

=== Binder Integration

When the form is backed by a [classname]`Binder`, pass the binder to the controller as well:

[source,java]
----
Binder<Employee> binder = new Binder<>(Employee.class);
binder.bindInstanceFields(this);

FormAIController controller = new FormAIController(form, binder);
----

> For every named binding (`bind("propertyName")`, `bindInstanceFields(this)`, or `@PropertyId`), the bean property name is used as a default field description, so when the user mentions a field by its bean-side name, the LLM can match the request to the right field. An explicit [methodname]`describe()` call always overrides the default. Lambda-bound bindings have no property name and contribute no default.

Reviewer 2: Does "contribute no default" mean "contribute no default field description"?

Author: Yes, added it explicitly.

> == Options for Select Fields

Reviewer 2: ```suggestion
== Options for Selection Fields
```

> For combo boxes, select fields, and multi-select components whose option set comes from the application, rather than a fixed enum on the field, register the options with [classname]`ValueOptions`:

Reviewer 2: ```suggestion
For [classname]`ComboBox`, [classname]`Select`, and multi-select components like [classname]`CheckboxGroup` whose option set comes from the application rather than a fixed enum on the field, register the options with [classname]`ValueOptions`:
```

[NOTE]
Single- and multi-select fields configured with [methodname]`setItems(...)` already share their items with the LLM, so the simple fixed-options case often needs no [methodname]`valueOptions()` call. Use [methodname]`valueOptions()` when items come from a lazy or remote source rather than an in-memory list, or when you want the LLM to fetch options through a filter callback instead of receiving the full set up front.

=== Multi-Select Fields

[classname]`MultiSelectComboBox`, [classname]`CheckboxGroup`, and any other field that implements [classname]`MultiSelect` are supported. Use the component's concrete multi-select type for the field reference so the [methodname]`forField(MultiSelect)` overload is selected:

[source,java]
----
MultiSelectComboBox<Project> projectsField = new MultiSelectComboBox<>("Projects");

controller.valueOptions(
        ValueOptions.forField(projectsField)
                .options(List.of("Apollo", "Vega", "Helios")),
        label -> projectService.findByName(label));
----

The converter runs once per chosen label, and the resolved values are written to the field as a set.

.Multi-Value Fields Must Implement MultiSelect
[NOTE]

> A field whose value type is a [classname]`Collection` must implement [classname]`MultiSelect`. The controller rejects two cases at registration time: a [classname]`MultiSelect` field passed through the single-value [methodname]`forField(HasValue)` overload, and a [classname]`Collection`-valued field that doesn't implement [classname]`MultiSelect`.

Reviewer 2: > a [classname]`MultiSelect` field passed through the single-value [methodname]`forField(HasValue)` overload

How can that even happen, if the signatures are identical?

Author: Java picks by the declared type, so a `MultiSelect` accessed through a `HasValue` (for example casting) reference will go through the single-value overload.

---
title: AI Form Filler
description: Use FormAIController to let users fill any Vaadin form from natural-language input or attached files.
meta-description: Configure the Vaadin FormAIController to fill form fields from natural language or attached files, with Binder validation and select-options support.
order: 80
---

> = [since:com.vaadin:vaadin@V25.2]#AI-Powered Form#

Reviewer 2: ```suggestion
= [since:com.vaadin:vaadin@V25.2]#AI Form Filler#
```
